torchvision/prototype/models/depth/stereo/raft_stereo.py

```diff
@@ -0,0 +1,711 @@
+        block2 = block(out_channels, out_channels, norm_layer=nn.BatchNorm2d, stride=1)
+        return nn.Sequential(block1, block2)
+
+    def forward(self, x: Tensor) -> List[Tensor]:
+        x = self.base_encoder(x)
+        outs = []
+        for layer_dict in self.downsample_and_out_layers:
+            x = layer_dict["downsampler"](x)
+            outs.append(torch.cat([layer_dict["out_hidden_state"](x), layer_dict["out_context"](x)], dim=1))
+        return outs
+
+
+class ConvGRU(raft.ConvGRU):
+    """Convolutional Gru unit."""
+
+    # Modified from raft.ConvGRU to accept pre-convolved contexts,
+    # see: https://github.com/princeton-vl/RAFT-Stereo/blob/main/core/update.py#L23
+    def forward(self, h: Tensor, x: Tensor, context: List[Tensor]) -> Tensor:  # type: ignore[override]
+        hx = torch.cat([h, x], dim=1)
+        z = torch.sigmoid(self.convz(hx) + context[0])
+        r = torch.sigmoid(self.convr(hx) + context[1])
+        q = torch.tanh(self.convq(torch.cat([r * h, x], dim=1)) + context[2])
+        h = (1 - z) * h + z * q
+        return h
+
+
+class MultiLevelUpdateBlock(nn.Module):
+    """The update block which contains the motion encoder and grus
+
+    It must expose a ``hidden_dims`` attribute which is the hidden dimension size of its gru blocks
+    """
+
+    def __init__(self, *, motion_encoder: MotionEncoder, hidden_dims: List[int]):
+        super().__init__()
+        self.motion_encoder = motion_encoder
+
+        # The GRU input size is the size of previous level hidden_dim plus next level hidden_dim
+        # if this is the first gru, then we replace previous level with motion_encoder output channels
+        # for the last GRU, we dont add the next level hidden_dim
+        gru_input_dims = []
+        for i in range(len(hidden_dims)):
+            input_dim = hidden_dims[i - 1] if i > 0 else motion_encoder.out_channels
+            if i < len(hidden_dims) - 1:
+                input_dim += hidden_dims[i + 1]
+            gru_input_dims.append(input_dim)
+
+        self.grus = nn.ModuleList(
+            [
+                ConvGRU(input_size=gru_input_dims[i], hidden_size=hidden_dims[i], kernel_size=3, padding=1)
+                # Ideally we should reverse the direction during forward to use the gru with smallest resolution first
+                # however currently there is no way to reverse a ModuleList that is jit script compatible
+                # hence we reverse the ordering of self.grus on the constructor instead
+                # see: https://github.com/pytorch/pytorch/issues/31772
+                for i in reversed(list(range(len(hidden_dims))))
+            ]
+        )
+
+        self.hidden_dims = hidden_dims
+
+    def _downsample2x(self, x):
+        return F.avg_pool2d(x, 3, stride=2, padding=1)
+
+    def _upsample2x(self, x):
+        _, _, h, w = x.shape
+        return F.interpolate(x, size=(2 * h, 2 * w), mode="bilinear", align_corners=True)
```

Comment:
Nit: I still think these 2 methods need to be removed, if not for readability, for ensuring we follow the best practices of other parts of the code-base. Usually introducing private methods used once is discouraged.